interop/nkic/gather.c

```diff
@@ -0,0 +1,85 @@
+/*
+Copyright (c) 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+*/
+#include "frontend.h"
+#include "ast_python.h"
+#include "ast_python_core.h"
+// TODO: only for testing
+#include "ast_nki.h"
+
+// Note: modern Pythons have PyImport_GetModuleByDef, but we need to be
+// compatible with 3.9
+
+static PyObject* get_util(const char *name) {
+  PyObject *m = NULL, *attr = NULL, *f = NULL;
+  PyObject *fe = PyUnicode_FromString("frontend");
+  if (fe) {
+    m = PyImport_GetModule(fe);
+    if (m) {
+      attr = PyUnicode_FromString(name);
+      if (attr)
+        f = PyObject_GetAttr(m, attr);
+    }
+  }
+  Py_XDECREF(fe);
+  Py_XDECREF(m);
+  Py_XDECREF(attr);
+  return f;
+}
+
+static struct _mod* parse_function(PyObject *f) {
+  struct _mod *m = NULL; // needed for done label
+
+  PyObject *getsrc = get_util("_get_src");
+  if (!getsrc)
+    return NULL;
+
+  PyObject *args = PyTuple_New(1);
+  if (!args) {
+    Py_DECREF(getsrc);
+    return NULL;
+  }
+
+  PyTuple_SetItem(args, 0, f);
```

Comment:
1) important: [PyTuple_SetItem()](https://docs.python.org/3/c-api/tuple.html#c.PyTuple_SetItem) "steals" a reference.
2) trivial: use [PyTuple_SET_ITEM()](https://docs.python.org/3/c-api/tuple.html#c.PyTuple_SET_ITEM). You're right that `PyTuple_SetItem()` can't fail in this case, so it's not worth checking the return code, and for that exact reason there's another version of this function that doesn't have a return code (but asserts on misuse)

```suggestion
  Py_INCREF(f);
  PyTuple_SET_ITEM(args, 0, f); /* steals reference to item */
```

Reply:
Good catch, I changed this.

We should be careful about relying on assert though. These are disabled in production builds of python.